// cdk-assets/lib/private/handlers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeAssetHandler = makeAssetHandler;
exports.destinationToClientOptions = destinationToClientOptions;
const container_images_1 = require("./container-images");
const files_1 = require("./files");
const asset_manifest_1 = require("../../asset-manifest");
function makeAssetHandler(manifest, asset, host, options) {
    if (asset instanceof asset_manifest_1.FileManifestEntry) {
        return new files_1.FileAssetHandler(manifest.directory, asset, host);
    }
    if (asset instanceof asset_manifest_1.DockerImageManifestEntry) {
        return new container_images_1.ContainerImageAssetHandler(manifest.directory, asset, host, options);
    }
    throw new Error(`Unrecognized asset type: '${asset}'`);
}
function destinationToClientOptions(destination) {
    // Explicitly build ClientOptions from AwsDestination. The fact they are structurally compatible is coincidental.
    // This also enforces better type checking that cdk-assets depends on the appropriate version of
    // @aws-cdk/cloud-assembly-schema.
    return {
        assumeRoleArn: destination.assumeRoleArn,
        assumeRoleExternalId: destination.assumeRoleExternalId,
        assumeRoleAdditionalOptions: destination.assumeRoleAdditionalOptions,
        region: destination.region,
    };
}
