// projen/lib/cli/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliError = void 0;
exports.installPackage = installPackage;
exports.renderInstallCommand = renderInstallCommand;
exports.findJsiiFilePath = findJsiiFilePath;
const fs = require("fs");
const path = require("path");
const logging = require("../logging");
const util_1 = require("../util");
/**
 * Installs the npm module (through `npm install`) to node_modules under `projectDir`.
 * @param spec The npm package spec (e.g. `foo@^1.2` or `foo@/var/folders/8k/qcw0ls5pv_ph0000gn/T/projen-RYurCw/pkg.tgz`)
 * @returns The installed package name (e.g. `@foo/bar`)
 */
function installPackage(baseDir, spec, isProjen = false) {
    const packageJsonPath = path.join(baseDir, "package.json");
    const packageJsonExisted = fs.existsSync(packageJsonPath);
    if (!packageJsonExisted) {
        // Make sure we have a package.json to read from later
        (0, util_1.exec)("npm init --yes", { cwd: baseDir });
    }
    logging.info(`installing module ${spec}...`);
    (0, util_1.exec)(renderInstallCommand(baseDir, spec), { cwd: baseDir });
    // Get the true installed package name
    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
    const packageName = Object.keys(packageJson.devDependencies).find((name) => isProjen ? name === "projen" : name !== "projen");
    if (!packageName) {
        throw new Error(`Unable to resolve package name from spec ${spec}`);
    }
    // if package.json did not exist before calling `npm install`, we should remove it
    // so we can start off clean.
    if (!packageJsonExisted) {
        fs.rmSync(packageJsonPath, { force: true, recursive: true });
    }
    return packageName;
}
/**
 * Render a command to install an npm package.
 *
 * Engine checks are ignored at this point so that the module can be installed
 * regardless of the environment. This was needed to unblock the upgrade of the
 * minimum node version of projen, but also okay generally because engine checks
 * will be performed later and for all eternity.
 *
 * @param dir Base directory
 * @param module The module to install (e.g. foo@^1.2)
 * @returns The string that includes the install command ("npm install ...")
 */
function renderInstallCommand(dir, module) {
    return `npm install --save --save-dev -f --no-package-lock --prefix="${dir}" ${module}`;
}
function findJsiiFilePath(baseDir, moduleName) {
    try {
        return path.dirname(require.resolve(`${moduleName}/.jsii`, {
            paths: [baseDir],
        }));
    }
    catch (error) {
        if (error instanceof Error &&
            "code" in error &&
            error.code === "MODULE_NOT_FOUND") {
            // the provided module is not a jsii module
            return undefined;
        }
        else {
            // unexpected error, throw it
            throw error;
        }
    }
}
class CliError extends Error {
    constructor(...lines) {
        super(lines.join("\n"));
        this.name = "CliError";
    }
}
exports.CliError = CliError;
