// projen/lib/cleanup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FILE_MANIFEST = void 0;
exports.cleanup = cleanup;
const fs = require("fs");
const path = require("path");
const glob = require("glob");
const common_1 = require("./common");
const logging = require("./logging");
exports.FILE_MANIFEST = `${common_1.PROJEN_DIR}/files.json`;
function cleanup(dir, newFiles, exclude) {
    try {
        const manifestFiles = getFilesFromManifest(dir);
        if (manifestFiles.length > 0) {
            // Use `FILE_MANIFEST` to remove files that are no longer managed by projen
            removeFiles(findOrphanedFiles(dir, manifestFiles, newFiles));
        }
        else {
            // Remove all files managed by projen with legacy logic
            removeFiles(findGeneratedFiles(dir, exclude));
        }
    }
    catch (e) {
        logging.warn(`warning: failed to clean up generated files: ${e.stack}`);
    }
}
function removeFiles(files) {
    for (const file of files) {
        fs.rmSync(file, { force: true, recursive: true });
    }
}
function findGeneratedFiles(dir, exclude) {
    const ignore = [
        ...readGitIgnore(dir),
        "node_modules/**",
        ...exclude,
        ".git/**",
    ];
    const files = glob.sync("**", {
        ignore,
        cwd: dir,
        dot: true,
        nodir: true,
        absolute: true,
    });
    const generated = new Array();
    for (const file of files) {
        const contents = fs.readFileSync(file, "utf-8");
        if (contents.includes(common_1.PROJEN_MARKER)) {
            generated.push(file);
        }
    }
    return generated;
}
function findOrphanedFiles(dir, oldFiles, newFiles) {
    return oldFiles
        .filter((old) => !newFiles.includes(old))
        .map((f) => path.resolve(dir, f));
}
function getFilesFromManifest(dir) {
    try {
        const fileManifestPath = path.resolve(dir, exports.FILE_MANIFEST);
        if (fs.existsSync(fileManifestPath)) {
            const fileManifest = JSON.parse(fs.readFileSync(fileManifestPath, "utf-8"));
            if (fileManifest.files) {
                return fileManifest.files;
            }
        }
    }
    catch (e) {
        logging.warn(`warning: unable to get files to clean from file manifest: ${e.stack}`);
    }
    return [];
}
function readGitIgnore(dir) {
    const filepath = path.join(dir, ".gitignore");
    if (!fs.existsSync(filepath)) {
        return [];
    }
    return fs
        .readFileSync(filepath, "utf-8")
        .split("\n")
        .filter((x) => x?.trim() !== "")
        .filter((x) => !x.startsWith("#") && !x.startsWith("!"))
        .map((x) => x.replace(/^\//, "")) // remove "/" prefix
        .map((x) => `${x}\n${x}/**`)
        .join("\n")
        .split("\n");
}
