// projen/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.build = exports.cdk = exports.cdktf = exports.cdk8s = exports.awscdk = exports.release = exports.javascript = exports.typescript = exports.python = exports.java = exports.vscode = exports.circleci = exports.gitlab = exports.github = exports.web = void 0;
__exportStar(require("./compare"), exports);
__exportStar(require("./component"), exports);
__exportStar(require("./dev-env"), exports);
__exportStar(require("./dependencies"), exports);
__exportStar(require("./docker-compose"), exports);
__exportStar(require("./file"), exports);
__exportStar(require("./gitattributes"), exports);
__exportStar(require("./gitpod"), exports);
__exportStar(require("./runner-options"), exports);
__exportStar(require("./ignore-file"), exports);
__exportStar(require("./ini"), exports);
__exportStar(require("./json"), exports);
__exportStar(require("./json-patch"), exports);
__exportStar(require("./logger"), exports);
__exportStar(require("./license"), exports);
__exportStar(require("./makefile"), exports);
__exportStar(require("./object-file"), exports);
__exportStar(require("./option-hints"), exports);
__exportStar(require("./project"), exports);
__exportStar(require("./project-build"), exports);
__exportStar(require("./project-tree"), exports);
__exportStar(require("./projects"), exports);
__exportStar(require("./projenrc"), exports);
__exportStar(require("./projenrc-json"), exports);
__exportStar(require("./readme"), exports);
__exportStar(require("./renovatebot"), exports);
__exportStar(require("./sample-file"), exports);
__exportStar(require("./semver"), exports);
__exportStar(require("./source-code"), exports);
__exportStar(require("./task"), exports);
__exportStar(require("./tasks"), exports);
__exportStar(require("./task-model"), exports);
__exportStar(require("./task-runtime"), exports);
__exportStar(require("./testing"), exports);
__exportStar(require("./textfile"), exports);
__exportStar(require("./toml"), exports);
__exportStar(require("./version"), exports);
__exportStar(require("./yaml"), exports);
__exportStar(require("./xmlfile"), exports);
// export submodules
exports.web = require("./web");
exports.github = require("./github");
exports.gitlab = require("./gitlab");
exports.circleci = require("./circleci");
exports.vscode = require("./vscode");
exports.java = require("./java");
exports.python = require("./python");
exports.typescript = require("./typescript");
exports.javascript = require("./javascript");
exports.release = require("./release");
exports.awscdk = require("./awscdk");
exports.cdk8s = require("./cdk8s");
exports.cdktf = require("./cdktf");
exports.cdk = require("./cdk");
exports.build = require("./build");
