// projen/lib/logging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ICON = void 0;
exports.debug = debug;
exports.verbose = verbose;
exports.info = info;
exports.error = error;
exports.warn = warn;
exports.empty = empty;
exports.disable = disable;
exports.isEnabled = isEnabled;
const chalk = require("chalk");
const common_1 = require("./common");
exports.ICON = "👾"; //'✨';
// disable logs if running inside a test
let enabled = common_1.IS_TEST_RUN ? false : true;
function log(isError, color, ...text) {
    // errors are always emitted, even if logs are disabled
    if (!enabled && !isError) {
        return;
    }
    console.error(`${exports.ICON} ${color(...text)}`);
}
function debug(...text) {
    if (process.env.DEBUG) {
        log(false, chalk.gray, ...text);
    }
}
function verbose(...text) {
    log(false, chalk.white, ...text);
}
function info(...text) {
    log(false, chalk.cyan, ...text);
}
function error(...text) {
    log(true, chalk.red, ...text);
}
function warn(...text) {
    log(false, chalk.yellow, ...text);
}
function empty(count = 1) {
    console.log("\n".repeat(count - 1));
}
function disable() {
    enabled = false;
}
function isEnabled() {
    return enabled;
}
