// projen/lib/project-tree.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectTree = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const component_1 = require("./component");
const json_1 = require("./json");
class ProjectTree extends component_1.Component {
    constructor(project) {
        super(project, new.target.name);
        this.file = new json_1.JsonFile(project, ".projen/tree.json", {
            readonly: true,
            marker: false,
            obj: () => ({
                ["//"]: "Experimental. Expect frequent changes to the structure of this file.",
                ...renderConstruct(project),
            }),
        });
    }
}
exports.ProjectTree = ProjectTree;
_a = JSII_RTTI_SYMBOL_1;
ProjectTree[_a] = { fqn: "projen.ProjectTree", version: "0.88.0" };
function renderConstruct(construct) {
    return {
        [construct.node.id]: {
            metadata: Object.fromEntries(construct.node.metadata.map(({ type, data }) => [type, data])),
            nodes: construct.node.children.length
                ? construct.node.children.reduce((nodes, c) => ({
                    ...nodes,
                    ...renderConstruct(c),
                }), {})
                : undefined,
        },
    };
}
