// projen/lib/util/constructs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMPONENT_SYMBOL = exports.PROJECT_SYMBOL = void 0;
exports.tryFindClosest = tryFindClosest;
exports.findClosestProject = findClosestProject;
exports.isProject = isProject;
exports.isComponent = isComponent;
exports.tagAsProject = tagAsProject;
exports.tagAsComponent = tagAsComponent;
exports.PROJECT_SYMBOL = Symbol.for("projen.Project");
exports.COMPONENT_SYMBOL = Symbol.for("projen.Component");
/**
 * Create a function to find the closest construct matching a predicate
 * @param predicate
 * @returns A function to find the closest construct matching the predicate
 */
function tryFindClosest(predicate) {
    return (construct) => construct?.node.scopes.reverse().find(predicate);
}
/**
 * Create a function to find the closest construct matching a predicate
 * @param predicate
 * @returns A function to find the closest construct matching the predicate
 */
function findClosestProject(construct) {
    if (isComponent(construct)) {
        return construct.project;
    }
    const project = tryFindClosest(isProject)(construct);
    if (!project) {
        throw new Error(`${new.target.name} at '${construct.node.path}' must be created in the scope of a Project, but no Project was found`);
    }
    return project;
}
function isProject(x) {
    return x !== null && typeof x === "object" && exports.PROJECT_SYMBOL in x;
}
function isComponent(x) {
    return x !== null && typeof x === "object" && exports.COMPONENT_SYMBOL in x;
}
function tagAs(scope, tag) {
    Object.defineProperty(scope, tag, { value: true });
}
function tagAsProject(scope) {
    tagAs(scope, exports.PROJECT_SYMBOL);
}
function tagAsComponent(scope) {
    tagAs(scope, exports.COMPONENT_SYMBOL);
}
